execution/gethexec/executionengine.go

```diff
@@ -873,14 +875,18 @@ func (s *ExecutionEngine) digestMessageWithBlockMutex(num arbutil.MessageIndex,
 	}
 
 	startTime := time.Now()
-	if s.prefetchBlock && msgForPrefetch != nil {
-		go func() {
-			_, _, _, err := s.createBlockFromNextMessage(msgForPrefetch, true)
-			if err != nil {
-				return
-			}
-		}()
-	}
+
+	// TODO: IMPORTANT !!!
+	//  This will cause a tracer to be called multiple times for the same block concurrently
```

Comment:
This is probably not the only thing that'll cause tracer to be called multiple times.
In normal case we have no reorgs, but a node might reexecute the same block multiple times for ultiple reasons.

I didn't view geth side yet, but I think a better solution will be to:
* remove all changes to blockchain
* Pass a vm.Config to ProduceBlock which could enable tracing
* have a config that controls tracing and feed it to ExecutionEngine (it currently does not have a config.. have a look how we use configs for e.g. SyncMonitorConfig)
* Inside func (s *ExecutionEngine) createBlockFromNextMessage - pass to ProduceBlock a vm.Config that enables tracing if it was enabled in config and only if isMsgForPrefetch == false